### ISSUE TITLE
haddock is painfully slow on large library builds (aka agda)

### ISSUE THREAD

**Issue by cartazio** [haskell/haddock#484]
on the agda code base, haddock runs for quite some time! I've not yet dug into how much is due to ghc as a frontend or not, but notating it here for now. It may very well be that the ghc api is the real culprit, but some serious profiling will be needed to zoom in on the root of the matter.

**Comment by bgamari** [comment]
Indeed, Haddock performance could use some more eyes. Perhaps you'd like to do some quick profiling and see where the major pain points are?

**Comment by mpickering** [comment]
There is some profiling info in #462

**Comment by asr** [comment]
+1

**Comment by alexbiehl** [comment]
Just found this ticket: Let me describe the current status.

Currently haddock (<= 2.18.1) looks for modules transitively depend on modules with ´TemplateHaskell` in the module graph.

**Comment by andreasabel** [comment]
I see you did some work at #621.

**Comment by alexbiehl** [comment]
Sorry for the hiccup. I meant to say that also in this area @duog has improved the handling of `TemplateHaskell` which currently causes haddock to rebuild _every_ module in the module graph. With ghc-head it only considers modules transitively depending on TemplateHaskell for compilation.

**Comment by andreasabel** [comment]
There is still the problem that an efficient data structure is used at https://github.com/haskell/haddock/blame/master/haddock-api/src/Haddock/Backends/Hyperlinker/Ast.hs#L46.  Given the linear time for each lookup, you have the quadratic factor without needing anymore explanation. Here is the evidence: http://lpaste.net/147305

**Comment by alexbiehl** [comment]
Yes, we could use a more efficient structure there. @andreasabel could you give it a shot in a PR?

On the other hand: This profile used haddock-2.16.1. If you take a look at the cost centers you will see that most of the time is spent in `everything` and `combine` and not in `enrichToken` though:

```
        enrich                            Haddock.Backends.Hyperlinker.Ast      5702         211   24.5   18.2    52.0   47.3
         everything                       Haddock.Syb                           5723     7487674    7.0    4.9    27.5   29.1
          combine                         Haddock.Syb                           5724     5299211   20.6   24.2    20.6   24.2```

**Comment by alexbiehl** [comment]
I just found https://github.com/mpickering/haddock/commit/65d0bc620f6e104e555112c8223f472005845d84 

Could you do a profile with this patch included?

**Comment by andreasabel** [comment]
I tried to install various versions of haddock (including master), with no success at first.  The build instructions I added in PR #670 would have saved me quite some time.

I see that in the meantime, you did the fix already in #669.

**Comment by duog** [comment]
Are you passing --hyperlinked-source to haddock? IIRC this code path is only used with that flag?
